### PR TITLE
Adding IA ebook preview button

### LINKED ISSUE #1592: Add Previews to Open Library editions pages
There are a lot of OL books we don't have -- but maybe we can at least provide google previews for these editions?

### PR DESCRIPTION
**Pull request by mekarpeles**
### Description

Closes #1592

preview
preview-modal

### CODE REVIEW

File: openlibrary/macros/LoanStatus.html [hunk 1]
```diff
@@ -75,29 +76,42 @@
...
 
-$elif page.get('ocaid') and not page.is_access_restricted() and editions_page:
+$elif ocaid and not page.is_access_restricted() and editions_page:
     $ viewbook = "//%s/stream/XXX?ref=ol" % bookreader_host()
-    <a href="$viewbook.replace('XXX', page.ocaid)" title="Use BookReader to read online" class="cta-btn cta-btn--available">Read</a>
+    <a href="$viewbook.replace('XXX', ocaid)" title="Use BookReader to read online" class="cta-btn cta-btn--available">Read</a>
 
-$elif (not page.get('ocaid') or page.is_access_restricted()) and editions_page:
+$elif (not ocaid or page.is_access_restricted()) and editions_page:
+  $ sponsorship = qualifies_for_sponsorship(page)
```
Comment (mekarpeles) [inline openlibrary/macros/LoanStatus.html]: Ack, this shouldn't be on this branch!

File: openlibrary/macros/LoanStatus.html [hunk 2]
```diff
@@ -75,29 +76,42 @@
...
+  $if (input(sponsorship=None) or user and user.is_sponsor()) and sponsorship.get('is_eligible'):
```
Comment (mekarpeles) [inline openlibrary/macros/LoanStatus.html]: Also should not be on this branch

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 3]
```diff
@@ -338,6 +338,24 @@ export default function init(){
             $('.wmd-preview').before('<h3 id="prevHead" style="margin:15px 0 10px;padding:0;">Preview</h3>');
         }
     });
+
+    // For Book Sponsorship Button
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/ol.js]: Comment is wrong, should be for book preview

File: static/css/components/buttonCta.less [hunk 4]
```diff
@@ -26,6 +26,22 @@ a.cta-btn {
...
+    color: @white;
+  }
+
+  &--preview, &--preview:link, &--preview:visited {
+    background-color: @white;
+    border: 1px solid @lighter-grey;
+    color: @black;
+  }
+
+  &--sponsor {
```
Comment (mekarpeles) [inline static/css/components/buttonCta.less]: `sponsor` should not be in this branch

**Comment by mekarpeles** [comment]
This is ready for review into `master-prerelease` but will need BK's blessing and a refactor of bookreader /embed/ view (to be discussed tomorrow/Friday) before it goes live

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 5]
```diff
@@ -338,6 +338,24 @@ export default function init(){
             $('.wmd-preview').before('<h3 id="prevHead" style="margin:15px 0 10px;padding:0;">Preview</h3>');
         }
     });
+
+    // For Book Preview Button
+    $(document).ready(function(){
```
Comment (jdlrobson) [inline openlibrary/plugins/openlibrary/js/ol.js]: No need for using .ready here.

Can this change wait for
https://github.com/internetarchive/openlibrary/pull/2305/commits/cf6527e30ec3b6afe9a923ca93e01218dce0bdca which is blocked on code review on https://github.com/internetarchive/openlibrary/pull/2344 

If those patches can be prioritised this will save me a lot of work in future and we can restrict this code to only the pages that need it.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 6]
```diff
@@ -338,6 +338,24 @@ export default function init(){
...
+        $('#bookPreviewButton').colorbox({
+            inline: true,
+            opacity: '0.5',
+            href: '#bookPreview'
+        })
+    })
+
+    $('.lazyIframe').show(function(){
```
Comment (jdlrobson) [inline openlibrary/plugins/openlibrary/js/ol.js]: https://api.jquery.com/show/ doesn't take a function as a parameter.  Chain instead?

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 7]
```diff
@@ -338,6 +338,24 @@ export default function init(){
...
+        // Find the iframes within our newly-visible element
+        $(this).find('iframe').prop('src', function(){
+            // Set their src attribute to the value of data-src
+            return $(this).data('src');
+        });
+    });
+
     initReadingListFeature();
     initBorrowAndReadLinks();
 }
```
Comment (jdlrobson) [inline openlibrary/plugins/openlibrary/js/ol.js]: Please don't put these changes in ol.js - break them out into a separate file called previewButton and check for the existence of #bookPreviewButton in the page before executing.

index.js should look something like this:

```
if ( $( '#bookPreviewButton' ).length ) {
   require( './bookPreview' )();
}
```

File: static/css/components/buttonCta.less [hunk 8]
```diff
@@ -6,26 +6,36 @@
 // override default anchor styles
 a.cta-btn {
   font-size: 14px;
-  color: @white !important;
```
Comment (jdlrobson) [inline static/css/components/buttonCta.less]: Are you sure you want to drop important here? Can you be sure you haven't broken any of the buttons that use cta-btn but try to override its color?

File: static/css/components/buttonCta.less [hunk 9]
```diff
@@ -6,26 +6,36 @@
...
+    color: @white;
+    text-decoration: none;
+  }
   &--available {
     background-color: @primary-blue;
   }
   &--unavailable {
     background-color: @orange;
   }
+  &--preview, &--preview:link, &--preview:visited {
```
Comment (jdlrobson) [inline static/css/components/buttonCta.less]: Do we really need to add another button to our long list? Can we not use one of the existing ones?

File: static/css/page-plain.less [hunk 10]
```diff
@@ -24,7 +24,6 @@ body#plain {
   /* stylelint-disable selector-max-specificity */
   #contentBody,
   #contentHead {
-    width: 690px;
```
Comment (jdlrobson) [inline static/css/page-plain.less]: I like it, but are you  sure this doesn't break anything?

File: openlibrary/macros/WorldcatLink.html [hunk 11]
```diff
@@ -3,10 +3,7 @@
 $if oclc_numbers or isbn:
   <hr>
   <div class="cta-section">
-    <p>
-      <span class="cta-section-title">Prefer the physical book?</span>
-      Check nearby libraries with:
-    </p>
+    <p class="cta-section-title">Check your local library</p>
```
Comment (mekarpeles) [inline openlibrary/macros/WorldcatLink.html]: Minor change, not moving to another PR

**Comment by cdrini** [comment]
@mekarpeles have you resolved the comments from your auto-cr that are fixed?

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 12]
```diff
@@ -338,6 +338,24 @@ export default function init(){
...
+        $('#bookPreviewButton').colorbox({
+            inline: true,
+            opacity: '0.5',
+            href: '#bookPreview'
+        })
+    })
+
+    $('.lazyIframe').show(function(){
+        // Find the iframes within our newly-visible element
+        $(this).find('iframe').prop('src', function(){
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/ol.js]: Store the iframe in a var; then this becomes easier to read:

```js
const $iframe = $(this).find('iframe');
$iframe.prop('src', $iframe.data('src'));
```

**Comment by cdrini** [comment]
@mekarpeles I think the UI you have is fine as is, but what do you think about this:
image

Changes:
- blue border on preview button to make it more visually consistent with big borrow button
- consumes less vertical space
- Daisy in same section as borrow (avoid another cta section)

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 6 above)
Comment (mekarpeles) [reply openlibrary/plugins/openlibrary/js/ol.js]: Really ?? Almost all the examples I see in those very docs take a function() as a parameter. Happy to change it if necessary

File: openlibrary/macros/LoanStatus.html [hunk 13]
```diff
@@ -75,29 +76,29 @@
...
-          <div class="check-availability">
-            <p>
-              Other editions of this book may be available:
-            </p>
-            <p class="check-other-editions">
-              <a data-ol-link-track="EditionPageCheckOtherEditionsButton" href="$work.url()">Check Other Editions</a>
-            </p>
-          </div>
+$if ocaid:
+  <p>
```
[the remark this answers is not in the document]
Comment (mekarpeles) [reply openlibrary/macros/LoanStatus.html]: That's true, we could probably get around it (by adding a different style), given this isn't a breaking change, I'd rather keep it for now (I don't want to worry about generalizing the CSS for the case where there are multiple buttons, even though it probably wouldn't be that hard)

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 12 above)
Comment (mekarpeles) [reply openlibrary/plugins/openlibrary/js/ol.js]: good suggestion, added

**Comment by mekarpeles** [comment]
@cdrini border color made blue, seems like a good change, similar theme for upcoming sponsorship btn as well

File: openlibrary/macros/LoanStatus.html [hunk 14]
```diff
@@ -52,52 +57,58 @@
...
+$if ocaid and page.is_access_restricted() and availability_status.startswith('borrow') and editions_page:
+  <p>
+    <a class="cta-btn cta-btn--preview" href="#bookPreview" id="bookPreviewButton">Preview</a>
+  </p>
+  <div class="hidden">
+    <div class="floater" id="bookPreview">
+      <div class="book-preview">
+        <div class="floaterHead">
+          <h2>$_('Preview Book')</h2>
+          <a class="floaterShut" data-key="book_preview" data-ol-link-track="book_preview"
```
Comment (cdrini) [inline openlibrary/macros/LoanStatus.html]: Add a `title=$_('Close')`

File: openlibrary/macros/LoanStatus.html [hunk 15]
```diff
@@ -52,52 +57,58 @@
...
+             href="javascript:;" onclick="\$.fn.colorbox.close();">&times;<span class="shift">$_("Close")</span></a>
```
Comment (cdrini) [inline openlibrary/macros/LoanStatus.html]: QUESTION: What's with the hidden span? Is that an accessibility hack?

File: openlibrary/macros/LoanStatus.html [hunk 16]
```diff
@@ -52,52 +57,58 @@
...
+        </div>
+        <div id="bookPreviewIframe" class="lazyIframe">
+          <iframe width="100%" height="515"
+                  data-src="https://archive.org/stream/$ocaid?wrapper=false">
+          </iframe>
+        </div>
+        <p class="learn-more">
+          <a href="https://archive.org/details/$ocaid" data-key="book_preview_learn_more"
+             data-ol-link-track="book_preview_learn_more">See more</a>
```
Comment (cdrini) [inline openlibrary/macros/LoanStatus.html]: I think these should have a hierarchy.
```suggestion
             data-ol-link-track="BookPreview|ArchiveLink">See more</a>
```

Comment (cdrini) [inline openlibrary/macros/LoanStatus.html]: Put everything inside the `<a>` for i18n.

File: openlibrary/macros/daisy.html [hunk 17]
```diff
@@ -1,19 +1,18 @@
-$def with(page, url='', msg='', block_name=None)
+$def with(page, url='', enc=False, block_name=None)
```
Comment (cdrini) [inline openlibrary/macros/daisy.html]: I think dropping `msg` breaks the search page.

Comment (cdrini) [inline openlibrary/macros/daisy.html]: Give this a longer name. Right now my only guess is `encoding`? Which doesn't make sense here.

File: openlibrary/plugins/openlibrary/js/ol.js [hunk 18]
```diff
@@ -417,3 +420,25 @@ export function initBorrowAndReadLinks() {
 
     /* eslint-enable no-unused-vars */
 }
+
+export function initPreviewButton() {
+    /**
+     * Colorbox modal + iframe for Book Preview Button
+     */
+    $(document).ready(function(){
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/ol.js]: You no longer need to call `$(document).read(`; just put the stuff in the function here directly.

File: static/css/components/cbox.less [hunk 19]
```diff
@@ -134,7 +134,6 @@ div.coverFloat {
 div.floater {
   position: relative;
   font-family: @lucida_sans_serif-1;
-  width: 640px;
```
Comment (cdrini) [inline static/css/components/cbox.less]: This breaks the `manage covers` modal (or the other change in this file, not sure)

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 6 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/ol.js]: I think changing it would best; since it's not one of the signatures on the API page, it could be deprecated without warning (it was also hard for me to figure out what this does cause I didn't see its signature on the API page, so readability improves).

**Comment by cdrini** [comment]
@mekarpeles Could we try the DONE/WON'T FIX pattern in the CONTRIBUTING?

> 3. **Resolve all code review (CR) comments**: Treat comments as a todo list. Most PRs will require some edits before getting merged, so don't get discouraged if you have to make some changes!
> 4. **Reply when resolving CR comments**: When resolving a comment, reply with either "DONE" or "WON'T FIX because ...". A reviewer will unresolve a comment if they feel it's necessary.

There are a number of unresolved comments whose status I'm unsure of (i.e. I don't if you've done them, or decided not to do them). (Labelling as "WON'T FIX because..." is a perfectly reasonable response :P)

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 7 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/ol.js]: The discussion is done, but I think it might be best to "grandfather policy" this PR since the changes were happening simultaneously. Future PRs will have to follow that pattern :+1:

File: static/css/components/buttonCta.less (hunk 8 above)
Comment (cdrini) [reply static/css/components/buttonCta.less]: Were these checked, @mekarpeles ?

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 6 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/ol.js]: 🤦‍♂ Nevermind! You're completely right, it's the second example!
image

I was misreading the `[]`. This is good as is :+1:

File: static/css/components/buttonCta.less (hunk 8 above)
Comment (mekarpeles) [reply static/css/components/buttonCta.less]: Yes, there's also a follow pr with my better world books changes which addresses other button discrepancies. I feel good about this  as DONE

File: static/css/page-plain.less (hunk 10 above)
Comment (mekarpeles) [reply static/css/page-plain.less]: Yes, tested

File: openlibrary/macros/LoanStatus.html [hunk 20]
```diff
@@ -75,29 +76,29 @@
...
-        $if work and work.edition_count > 1:
-          <div class="check-availability">
-            <p>
-              Other editions of this book may be available:
-            </p>
-            <p class="check-other-editions">
-              <a data-ol-link-track="EditionPageCheckOtherEditionsButton" href="$work.url()">Check Other Editions</a>
-            </p>
-          </div>
+$if ocaid:
```
[the remark this answers is not in the document]
Comment (mekarpeles) [reply openlibrary/macros/LoanStatus.html]: It should probably just be a macro and not need a comment. Will make this change

File: openlibrary/macros/daisy.html (hunk 17 above)
Comment (mekarpeles) [reply openlibrary/macros/daisy.html]: Thank you, this is a BLOCKER!

File: static/css/components/cbox.less (hunk 19 above)
Comment (mekarpeles) [reply static/css/components/cbox.less]: Will investigate

File: static/css/components/buttonCta.less (hunk 9 above)
Comment (mekarpeles) [reply static/css/components/buttonCta.less]: One option is to have a generic secondary call to action which can be used both for the preview and the sponsorship button.

File: openlibrary/macros/daisy.html (hunk 17 above)
Comment (mekarpeles) [reply openlibrary/macros/daisy.html]: Let's call this `protected`

File: openlibrary/macros/LoanStatus.html (hunk 14 above)
Comment (mekarpeles) [reply openlibrary/macros/LoanStatus.html]: incoming

File: openlibrary/macros/LoanStatus.html (hunk 16 above)
Comment (mekarpeles) [reply openlibrary/macros/LoanStatus.html]: done

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 18 above)
Comment (mekarpeles) [reply openlibrary/plugins/openlibrary/js/ol.js]: done

File: openlibrary/plugins/openlibrary/js/ol.js (hunk 5 above)
Comment (mekarpeles) [reply openlibrary/plugins/openlibrary/js/ol.js]: done

File: static/css/components/buttonCta.less (hunk 9 above)
Comment (mekarpeles) [reply static/css/components/buttonCta.less]: punting for now, will not fix

File: openlibrary/macros/LoanStatus.html (hunk 20 above)
[the remark this answers is not in the document]
Comment (mekarpeles) [reply openlibrary/macros/LoanStatus.html]: moved to macro, done

File: openlibrary/macros/LoanStatus.html (hunk 16 above)
Comment (mekarpeles) [reply openlibrary/macros/LoanStatus.html]: punting for now, will not fix

File: openlibrary/macros/daisy.html (hunk 17 above)
Comment (mekarpeles) [reply openlibrary/macros/daisy.html]: done

File: static/css/components/cbox.less (hunk 19 above)
Comment (mekarpeles) [reply static/css/components/cbox.less]: fixed issue by adding args to colorbox init, not by adding css

File: openlibrary/macros/WorldcatLink.html (hunk 11 above)
Comment (mekarpeles) [inline openlibrary/macros/WorldcatLink.html]: This should probably be i18n'ified

Comment (tfmorris) [reply openlibrary/macros/WorldcatLink.html]: Yes, all text needs to be translatable.
